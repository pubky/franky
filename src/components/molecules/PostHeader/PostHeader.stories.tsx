import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { PostHeader } from './PostHeader';

const meta: Meta<typeof PostHeader> = {
  title: 'Molecules/Post/PostHeader',
  component: PostHeader,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
      values: [
        { name: 'dark', value: '#05050A' },
        { name: 'light', value: '#F2F2F7' },
      ],
    },
  },
  args: {
    avatarSrc: '/images/avatar.png',
    displayName: 'This is a text',
    label: 'Label',
    timeLabel: '15m',
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = { render: (args) => <PostHeader {...args} /> };
