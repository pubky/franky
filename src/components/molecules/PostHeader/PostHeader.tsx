import * as React from 'react';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

export interface PostHeaderProps {
  avatarSrc?: string;
  avatarAlt?: string;
  displayName: string;
  label?: string;
  timeLabel?: string;
  className?: string;
}

export function PostHeader({
  avatarSrc,
  avatarAlt = 'avatar',
  displayName,
  label,
  timeLabel,
  className,
}: PostHeaderProps) {
  return (
    <div className={Libs.cn('flex items-start justify-between w-full', className)}>
      <div className="flex items-center gap-3">
        <Atoms.Avatar size="lg">
          <Atoms.AvatarImage src={avatarSrc} alt={avatarAlt} />
          <Atoms.AvatarFallback>{Libs.extractInitials({ name: displayName, maxLength: 2 })}</Atoms.AvatarFallback>
        </Atoms.Avatar>
        <div className="flex flex-col">
          <span className="text-base leading-6 font-bold text-foreground">{displayName}</span>
          {label && (
            <span className="text-xs leading-4 font-medium tracking-[0.075em] uppercase text-muted-foreground">
              {label}
            </span>
          )}
        </div>
      </div>
      {timeLabel && (
        <div className="flex items-center gap-1 text-xs font-medium tracking-[0.075em] text-muted-foreground">
          <Libs.Clock className="size-4 text-muted-foreground" />
          <span>{timeLabel}</span>
        </div>
      )}
    </div>
  );
}
