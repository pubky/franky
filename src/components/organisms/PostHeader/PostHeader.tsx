'use client';

import { useLiveQuery } from 'dexie-react-hooks';
import * as Core from '@/core';
import * as Libs from '@/libs';
import * as Atoms from '@/atoms';

export interface PostHeaderProps {
  postId: string;
  className?: string;
}

export function PostHeader({ postId, className }: PostHeaderProps) {
  const [userId] = postId.split(':');

  // Fetch post details to get indexed_at
  const postDetails = useLiveQuery(async () => {
    return await Core.PostDetailsModel.findById(postId);
  }, [postId]);

  // Fetch user details for avatar and name
  const userDetails = useLiveQuery(async () => {
    try {
      return await Core.ProfileController.read({ user_id: userId as Core.Pubky });
    } catch (error) {
      console.error('Error fetching user details:', error);
      return null;
    }
  }, [userId]);

  if (!postDetails || !userDetails) {
    // TODO: Add skeleton loading component for PostHeader
    return <div className="text-muted-foreground">Loading header...</div>;
  }

  return (
    <div className={Libs.cn('flex items-start justify-between w-full', className)}>
      <div className="flex items-center gap-3">
        <Atoms.Avatar size="lg">
          <Atoms.AvatarImage src={userDetails.image || undefined} />
          <Atoms.AvatarFallback>{Libs.extractInitials({ name: userDetails.name, maxLength: 2 })}</Atoms.AvatarFallback>
        </Atoms.Avatar>
        <div className="flex flex-col">
          <span className="text-base leading-6 font-bold text-foreground">{userDetails.name}</span>
          <span className="text-xs leading-4 font-medium tracking-[0.075em] uppercase text-muted-foreground">
            @{Libs.formatPublicKey({ key: userId, length: 8 })}
          </span>
        </div>
      </div>
      {Libs.timeAgo(new Date(postDetails.indexed_at)) && (
        <div className="flex items-center gap-1 text-xs font-medium tracking-[0.075em] text-muted-foreground">
          <Libs.Clock className="size-4 text-muted-foreground" />
          <span>{Libs.timeAgo(new Date(postDetails.indexed_at))}</span>
        </div>
      )}
    </div>
  );
}
