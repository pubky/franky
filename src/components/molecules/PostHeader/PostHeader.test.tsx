import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { PostHeader } from './PostHeader';

vi.mock('@/atoms', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/atoms')>();
  return {
    ...actual,
    Avatar: ({ src, alt }: { src?: string; alt?: string }) => (
      <div data-testid="avatar" data-src={src} aria-label={alt} />
    ),
  };
});

vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
    Clock: ({ className }: { className?: string }) => <div data-testid="clock" className={className} />,
  };
});

describe('PostHeader', () => {
  it('renders name, label and time', () => {
    render(<PostHeader avatarSrc="/a.png" displayName="Satoshi" label="label" timeLabel="15m" />);
    expect(screen.getByText('Satoshi')).toBeInTheDocument();
    expect(screen.getByText(/label/i)).toBeInTheDocument();
    expect(screen.getByText('15m')).toBeInTheDocument();
    expect(screen.getByTestId('avatar')).toBeInTheDocument();
    expect(screen.getByTestId('clock')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<PostHeader avatarSrc="/a.png" displayName="Satoshi" label="label" timeLabel="15m" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
