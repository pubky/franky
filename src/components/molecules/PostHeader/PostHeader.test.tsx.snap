// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostHeader > matches snapshot 1`] = `
<div
  class="flex items-start justify-between w-full"
>
  <div
    class="flex items-center gap-3"
  >
    <div
      data-testid="avatar"
    />
    <div
      class="flex flex-col"
    >
      <span
        class="text-base leading-6 font-bold text-foreground"
      >
        Satoshi
      </span>
      <span
        class="text-xs leading-4 font-medium tracking-[0.075em] uppercase text-muted-foreground"
      >
        label
      </span>
    </div>
  </div>
  <div
    class="flex items-center gap-1 text-xs font-medium tracking-[0.075em] text-muted-foreground"
  >
    <div
      class="size-4 text-muted-foreground"
      data-testid="clock"
    />
    <span>
      15m
    </span>
  </div>
</div>
`;
